qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteCodeActions.java

```diff
@@ -100,37 +114,115 @@ public CompletableFuture<List<CodeAction>> doCodeActions(Template template, Code
-					case UndefinedObject:
-						// The following Qute template:
-						// {undefinedObject}
-						//
-						// will provide a quickfix like:
-						//
-						// Declare `undefinedObject` with parameter declaration."
-						doCodeActionsForUndefinedObject(template, diagnostic, errorCode, codeActions);
-						break;
-					case UndefinedSectionTag:
-						// The following Qute template:
-						// {#undefinedTag }
-						//
-						// will provide a quickfix like:
-						//
-						// Create `undefinedTag`"
-						doCodeActionsForUndefinedSectionTag(template, diagnostic, codeActions);
-						break;
-					case UndefinedNamespace:
-						// The following Qute template:
-						// {undefinedNamespace:xyz}
-						doCodeActionToSetIgnoreSeverity(template, Collections.singletonList(diagnostic), errorCode,
-								codeActions, UNDEFINED_NAMESPACE_SEVERITY_SETTING);
-						break;
-					default:
-						break;
+						case UndefinedObject:
+							// The following Qute template:
+							// {undefinedObject}
+							//
+							// will provide a quickfix like:
+							//
+							// Declare `undefinedObject` with parameter declaration."
+							doCodeActionsForUndefinedObject(template, diagnostic, errorCode, codeActions);
+							break;
+						case UndefinedSectionTag:
+							// The following Qute template:
+							// {#undefinedTag }
+							//
+							// will provide a quickfix like:
+							//
+							// Create `undefinedTag`"
+							doCodeActionsForUndefinedSectionTag(template, diagnostic, codeActions);
+							break;
+						case UndefinedNamespace:
+							// The following Qute template:
+							// {undefinedNamespace:xyz}
+							doCodeActionToSetIgnoreSeverity(template, Collections.singletonList(diagnostic), errorCode,
+									codeActions, UNDEFINED_NAMESPACE_SEVERITY_SETTING);
+							break;
+						case UnknownProperty:
+							doCodeActionToCreateProperty(template, diagnostic, errorCode, codeActions,
+									codeActionResolveFutures,
+									resolver,
+									sharedSettings);
+							break;
+						default:
+							break;
 					}
 				}
 			}
 		}
-		return CompletableFuture.completedFuture(codeActions);
+		CompletableFuture<Void>[] registrationsArray = new CompletableFuture[codeActionResolveFutures.size()];
+		codeActionResolveFutures.toArray(registrationsArray);
+		return CompletableFuture.allOf(registrationsArray).thenApply((Void _void) -> {
+			return codeActions;
+		});
+	}
+
+	private void doCodeActionToCreateProperty(Template template, Diagnostic diagnostic, QuteErrorCode errorCode,
+			List<CodeAction> codeActions, List<CompletableFuture<Void>> registrations,
+			QuteTemplateGenerateMissingJavaMember resolver, SharedSettings settings) {
+		JsonObject data = (JsonObject) diagnostic.getData();
+		if (data == null) {
+			// TODO: the client may not support diagnostic data; resolve the needed
+			// information some other way
+			return;
+		}
+		MissingMemberData missingMemberData = DiagnosticDataFactory.getMissingMemberData(data);

```

Comment:
I think it should be better to have

UnkwonwPropertyData data = DiagnosticDataFactory.getUnkwonwPropertyData(diagnostic, template, javaCache);

In this method you could manage the case when diagnostic has a data and when diagnostic has no the data. In this case  node part must be found by the diagnostic index and must be resolved by using javaCache

Reply:
@AlexXuChen could use it this method too for similar code action.

Reply:
CodeActions don't have access to the java cache in this PR, so I didn't include that parameter

Reply:
Yes sure, let's do that in the PR which will support no diagnostic data. For the moment you can do:

`UnkwonwPropertyData data = DiagnosticDataFactory.getUnkwonwPropertyData(diagnostic);`